src/ApiCore/CallSettings.php

```diff
@@ -41,8 +41,8 @@ class CallSettings
 
     private $retrySettings;
     private $userHeaders;
-    private $grpcOptions;
-    private $restOptions;
+    private $transportOptions;
+    private $timeoutMillis;
```

Comment:
Looking at the changes, I am not sure what adding `$timeoutMillis` here is gaining us.

I understand that we want to provide an easy way for users to set the timeout in the call, so that is an argument for adding a `timeoutMillis` optional arg to the gapic methods. But once we have that, why don't we just take that user-provided arg and use it to set the `initialRpcTimoutMillis` and `noRetriesTimeoutMillis`, similar to what we were already doing in the `configureCallSettings()` method? That means we don't need to extra field in CallSettings and don't need the extra logic in the Retry middleware.

Reply:
The main reason is that the transport layers shouldn't need to look into retry settings. There is duplicate logic in handling the timeout in the transport layers.  If we use `initialRpcTimoutMillis` and `noRetriesTimeoutMillis` instead, then each transport has to look into the fields and do this, or something like it:

```php
$timeout = null;
if ($settings->retryEnabled()) {
    $timeout = $settings->getRetrySettings()->getInitialRpcTimeoutMillis();
} else {
    $timeout = $settings->getRetrySettings()->getNoRetriesTimeoutMillis();
}
```
Right now the transport does this:
```php
$timeout = $settings->getTimeoutMillis();
```

Additionally, `initialRpcTimeoutMillis` intuitively wouldn't change each retry.

Reply:
Got it, that explains why I am confused. I think my actual disagreement with what we are doing is focused somewhere else: we shouldn't have `CallSettings` be part of the signature of the callable returned here.

Reply:
Got it, that explains why I am confused. I think my actual disagreement with what we are doing is focused somewhere else: we shouldn't have `CallSettings` be part of the signature of the callable returned from here: https://github.com/googleapis/gax-php/pull/118/files#diff-499ff24bbcb2ebc2894a1138637631a5R193

The `CallSettings` object should be used to *construct* the callable, not be passed as an argument to the callable. There should be an options parameter of some sort (it could be an array, or a dedicated class called `TransportCallOptions` or something) but not `CallSettings`

Then the Retrying middleware would accept the `CallSettings` object in its constructor here: https://github.com/googleapis/gax-php/pull/118/files#diff-fe4a9f1b4b922c01aa282bd750c22646R149

And the action that the Retrying middleware would take is to modify the timeout in `TransportCallOptions` (or the array of options) and pass it down the stack.

Does that make sense?

Reply:
Yes it makes sense. What if we just separate `CallSettings` and `RetrySettings`?

```php
public function getCallable(RetrySettings $retrySettings)
{
    $callable = function (Call $call, CallSettings $settings) {
        // ...
    }
    // other middlewares added here
    return Middleware\RetryMiddleware($callable, $retrySettings);
}
```

And then in the GAPIC clients:

```php
public function startCall(Call $call, CallSettings $settings, RetrySettings, $retrySettings)
{
    $callable = $this->transport->getCallable($retrySettings);
    return $callable(
        $call,
        $settings
    );
}
```

Reply:
Yes, that works. But then should we pass `RetrySettings` directly, or should it be part of an $options array, or another class, perhaps `CallConstructionSettings`?

The reason I am pushing for that is because if we add a new setting/option that is required at callable construction time, we will need to pass it into `startCall`, but it doesn't make sense to add it to `CallSettings` if we are using `CallSettings` as the settings for *making* the call rather than *constructing* the call.

```
public function startCall(Call $call, CallSettings $settings, array $callConstructionOptions)
{
    // $callConstructionOptions contains 'retrySettings' => RetrySettings()'
    $callable = $this->transport->getCallable($callConstructionOptions);
    return $callable(
        $call,
        $settings
    );
}
```

Reply:
The delineation between settings for making the call and constructing the call make a lot of sense. Since the `Call` object already contains some information for constructing the call, what do you think about utilizing it for construction options like retries?

Reply:
I'm not a super fan of making it part of the `Call` object, personally. The `$callConstructionOptions` method above sounds good to me.

Reply:
I don't have strong feelings about the `Call` object, but it looks like the `Call` object is used when *making* the call as well. I think that whatever we choose, we should have a clean separation between settings used for construction and settings used for making the call.

I also think there is value in keeping the *fixed* values separate from the *changeable* values. Right now the `Call` object contains things that are totally fixed, so (1) perhaps the `Call` object should only be used at construction time, and (2) we should keep those fixed settings separate from the variable settings used at construction time, such as Retry settings.

Reply:
How are we defining making the call vs. constructing the call? This is what I have been thinking, please correct me if this is wrong:

[constructing the call](https://github.com/googleapis/gax-php/blob/b3bedf376d58c03ae4528877db9d2b99eb07b963/src/ApiCore/RestTransport.php#L111-L115)
[settings for making (or delivering?) the call](https://github.com/googleapis/gax-php/blob/b3bedf376d58c03ae4528877db9d2b99eb07b963/src/ApiCore/RestTransport.php#L116)

Reply:
I was categorizing somewhat differently. Things needed for constructing the call is everything in this method signature: https://github.com/googleapis/gax-php/blob/b3bedf376d58c03ae4528877db9d2b99eb07b963/src/ApiCore/RestTransport.php#L105
Things needed for making the call is everything in the callable signature: https://github.com/googleapis/gax-php/blob/b3bedf376d58c03ae4528877db9d2b99eb07b963/src/ApiCore/RestTransport.php#L107

Reply:
Thanks for the explanation. 

> Right now the Call object contains things that are totally fixed, so (1) perhaps the Call object should only be used at construction time

There is one exception to this in [Page.php](https://github.com/googleapis/gax-php/blob/30f9995614ee1481f21bb25c034fd5877721a701/src/ApiCore/Page.php#L123-L138), the message needs to be updated with the new page token. That bit right there is what actually inspired the decision to have the `Call` class in the first place.

Reply:
Additionally, I like the idea of the `Call` being something that could eventually be modified by users when they are able to inject their own middleware layers.  I could just be overestimating the usefulness of that, however.

Reply:
Hmm, I think I have been a bit confused (and confusing) about these distinctions. What about a different, clearer set of distinctions between the inputs:
  (1) Things that will ALWAYS be the same between calls (like $method, $decodeType in Call object)
  (2) Things that can change, but are NOT used by the transport layer (retries)
  (3) Things that can change, and ARE used by the transport layer ($message in Call object, timeout, headers).

I think that (1) and (2) should be passed to `getCallable`, and (3) should be in the signature of the callable.

Note also that the function of middlewares is to modify the things in bullet 3 before passing them on to the transport, e.g. retry middleware adjusts the timeout based on the multiplier, header middleware augments the headers. The Page object is not strictly middleware, but it does a similar thing by modifying the page token in the request. (Should it be middleware??)

Right now the `Call` object spans (1) and (3). I don't know what the right restructuring is, but I think we should separate those into different objects/args. I also like the idea of custom middleware. I think that bundling the request object, timeout and headers into a single object (`Call` or whatever we call it) without the fixed $method and $decodeType makes it easy to see what your middleware should be modifying. WDYT?

Reply:
>  I think that bundling the request object, timeout and headers into a single object (Call or whatever we call it) without the fixed $method and $decodeType makes it easy to see what your middleware should be modifying. WDYT?

I'm on board with that 👍

Reply:
Been thinking about this a bit. While items like `$method` and `$decodeType` may not need to be modified, is there harm in exposing them for conditional logic in middleware?

As a hypothetical (and not very practical) example,

What if someone would like to add a special header in the case of `$decodeType === SomeExpectedType::class`?